File: dp_fibonacci/dp_fibonacci.rb
def fib(n)
  case n
  when 0
    return 0
  when 1
    return 1
  else
    # F(n + 2) = F(n) + F(n + 1)
    # F = F(n - 2) + F(n - 1)
    return fib(n - 2) + fib(n - 1)
  end
end

def fib_dp(n)
  memo = [0, 1]
  i = 2
  while(n >= i) do
    memo[i] = memo[i - 1] + memo[i - 2]
    i = i + 1
  end
  memo[n]
end

File: dp_fibonacci/dp_fibonacci.go
package main

import (
  "fmt"
)

func fib(n int) int{
  switch n {
  case 0: return 0
  case 1: return 1
  default: return fib(n-2) + fib(n-1)
  }
}

func fib_dp(n int) int{
  memo := []int{0, 1}
  i := 2
  for n >= i {
    memo = append(memo, memo[i-1] + memo[i -2])
    i++
  }

  return memo[n]
}

func main() {
  fmt.Println("hello from dp")
}
